keps/sig-api-machinery/20190425-structural-openapi.md

```diff
@@ -0,0 +1,812 @@
+ 
+1. **structurality:** both `ForbiddenGenerics` and `ForbiddenExtensions` only have zero values, with the two exceptions of (4).
+2. **completeness:** the schema is _structurally complete_
+3. **RawExtension:** for every schema with `x-kubernetes-embedded-resource: true`, `x-kubernetes-preserve-unknown-fields: true` and `type: object` are set
+4. **IntOrString:** for `x-kubernetes-int-or-string: true` either `type` is empty under `anyOf` and `allOf` or the schema structure is either
+ 
+   ```yaml
+   x-kubernetes-int-or-string: true
+   anyOf:
+   - type: integer
+   - type: string
+   ...
+   ```
+   or
+   ```yaml
+   x-kubernetes-int-or-string: true
+   allOf:
+   - anyOf:
+     - type: integer
+     - type: string
+   - ...
+   ...
+   ```
+
+Structurally complete means that any specified JSON path in a schema has a specified type outside of any logical junctors (`andOf`,`anyOf`,`oneOf`,`not`; compare the guiding principle above and its example). In other words, we validate that 
+
+- every object field (`properties`), 
+- every array value (`items`) and 
+- every string map value (`additionalProperties`)
+
+has a non-empty `type` outside of logical junctors, with the possible exceptions of
+- `x-kubernetes-int-or-string: true` is set in the empty-type schema
+- `x-kubernetes-preserve-unknown-fields: true` is set in the empty-type schema.
+
+If such a `type` is empty, we report the following structural schema validation error (via the `NonStructuralSchema` condition messages, and via real API rejection according to the [API Compatibility Plan](#api-compatibility-plan)), e.g.:
+
+```
+.properties[foo].items.properties[bar].type must be non-empty
+```
+
+The additional validations have the following consequences:
+
+1. `additionalProperties`, `description`, `title`, `nullable`, `default`, `readOnly` and `type` (with the two `x-kubernetes-int-or-string: true` exceptions) and the new `x-kubernetes-*` extensions cannot be used inside of logical junctors `allOf`, `anyOf`, `oneOf`, `not`
+2. When setting `StructuralSchema.ValueValidatoin` recursively to nil in a schema, the schema becomes weaker.
+
+The second consequence means that we can correctly validate an object without understanding value validations (i.e. without rejecting an object as invalid which is actually valid). This is crucial for knowingly incomplete schema algorithms as used in pruning, defaulting, apply and elsewhere.
+
+### Unfolding the Extensions
+
+During publishing we unfold the new `x-kubernetes-*` extensions:
+
+* `x-kubernetes-int-or-string: true` unfolds to
+  ```yaml
+  x-kubernetes-int-or-string: true
+  anyOf:
+  - type: integer
+  - type: string
+  ```
+  if `ValueValidation.AnyOf` is unset, otherwise to: 
+  ```yaml
+  x-kubernetes-int-or-string: true
+  allOf:
+  - anyOf:
+    - type: integer
+    - type: string
+  - <ValueValidation.AnyOf value>
+  ```
+* `x-kubernetes-embedded-resource: true` unfolds to
```

Comment:
apiVersion and kind would be required, correct?

Reply:
oh right. adding.